### PR TITLE
House of arena

### PR DESCRIPTION
**Pull request by xmzyshypnc**
Hi guys, recently a new heap attack called house of husk is popular. I think the technique is too tricky for beginners. So I figured out a new method which is more universal. I name it as house-of-arena. This technique was used first on heap_master in \*CTF2019. I don't know if it's good enough to be called "house-of-*" but I just want to that :). Glad my last commit was accepted and still many thanks to you guys.
Best wishes~

### CODE REVIEW

**Comment by Kyle-Kyle** [comment]
Do you have a link to house of husk? I think I read it somewhere. Was it about printf functions?

**Comment by Kyle-Kyle** [comment]
I just read the technique. Basically, this is a combination of two techinques:
1. use unsorted-bin-attack to overwrite `global_max_fast` so we can write a heap address after `main_arena`(this technique is not new, it has been around for years. But we don't have it on how2heap)
2. use the trick of manipulating `fastbin` to write an arbitrary value onto `main_arena`. In this case, the `main_arena` is extended, so it is arbitrary value onto anywhere after `main_arena`. (this technique is not new either, the first time I saw it was a writeup of 0ctf2019. We don't have it on how2heap either.)

Personally, I would like to separate them into two different techniques.
@m1ghtym0 what do you say?

File: glibc_2.26/house_of_arena.c [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+#include <stdlib.h>
+#include <string.h>
+
+#define off2sz(off) ((off) * 2 - 0x10)
+
+#define MAIN_ARENA          0x3ebc40
+#define MAIN_ARENA_DELTA    0x60
+#define GLOBAL_MAX_FAST     0x3ed940
+#define FREE_HOOK           0x3ed8e8
+#define SYSTEM              0x4f440
```
[the remark this answers is not in the document]
Comment (xmzyshypnc) [reply glibc_2.26/house_of_arena.c:15]: Yeah I think you are right about this tech :) It’s a little tricky but still a combination of two single attack in essence. I treated it as a more universal tech than house-of-husk and ignored it’s essence. 
Thanks for your earnest  reply and careful review. Best wishes ^_^

[the remark this answers is not in the document]
Comment (xmzyshypnc) [reply glibc_2.26/house_of_arena.c:15]: Besides, your advice is really useful. I will pay more attention to my code and hope to make more contributions to this awesome repo lol

**Comment by m1ghtym0** [comment]
> I just read the technique. Basically, this is a combination of two techinques:
> 
> 1. use unsorted-bin-attack to overwrite `global_max_fast` so we can write a heap address after `main_arena`(this technique is not new, it has been around for years. But we don't have it on how2heap)
> 2. use the trick of manipulating `fastbin` to write an arbitrary value onto `main_arena`. In this case, the `main_arena` is extended, so it becomes an arbitrary 8-byte write onto anywhere after `main_arena`. (this technique is not new either, the first time I saw it was a writeup of 0ctf2019. We don't have it on how2heap either.)
> 
> Personally, I would like to separate them into two different techniques.
> @m1ghtym0 what do you say?

In general, I'd say that we should be careful not to be too inflationary with the `house_of_...`, so people are not overwhelmed and confused.
I understand that many of the new techniques are combining primitives present in other techniques to build new chains of heap interactions.
The goal of this repo is to cover many of theses primitives to educate people and give them the ability to be creative on their own. Imo, if you get the general concept of allocator implementations and see some of the houses here, writing arbitrary heap exploit comes naturally and is just putting the pieces together.
The `global_max_fast` trick seems to be one of those pieces, so why not add it to the repo?:-)

**Comment by mdulin2** [comment]
I wrote up something for the _House of Husk_. Hope this can clear things up and how things work: https://maxwelldulin.com/BlogPost?post=3107454976

**Pull request closed without merging**